### PR TITLE
Feature/type annotations and mypy

### PR DESCRIPTION
**Pull request by johnbywater**

### CODE REVIEW

File: eventsourcing/types.py [hunk 1]
```diff
@@ -0,0 +1,236 @@
...
+except ImportError:
+    # In Python 3.7, we can just do this:
+    class MetaAbstractDomainEntity(ABCMeta):
+        pass
+
+
+else:
+    # But in Python3.6 we must instead do this:
+    class MetaAbstractDomainEntity(GenericMeta):
+        pass
```
Comment (AlanFoster) [inline eventsourcing/types.py]: Just confirming if there's a semantic difference between the above, and something like:

```python
# Need to deal with the fact that Python3.6 had GenericMeta.
try:
    from typing import GenericMeta
    class MetaAbstractDomainEntity(GenericMeta):
        pass
except ImportError:
    # In Python 3.7, we can just do this:
    class MetaAbstractDomainEntity(ABCMeta):
        pass
```

File: eventsourcing/application/process.py [hunk 2]
```diff
@@ -378,6 +380,7 @@ def collect_pending_events(self, aggregates):
         num_changed_aggregates = 0
         # This doesn't necessarily obtain events in causal order...
         for aggregate in aggregates:
+            aggregate: BaseAggregateRoot
```
[the remark this answers is not in the document]
Comment (johnbywater) [reply eventsourcing/application/process.py]: Yes, that's true.

File: eventsourcing/domain/model/aggregate.py [hunk 3]
```diff
@@ -1,44 +1,46 @@
...
         """
         self.__pending_events__.append(event)
 
-    def __save__(self):
+    def __save__(self) -> None:
         """
         Publishes all pending events to subscribers.
         """
-        batch_of_events = self.__batch_pending_events__()
+        batch_of_events: List[AbstractDomainEvent] = self.__batch_pending_events__()
```
[the remark this answers is not in the document]
Comment (johnbywater) [reply eventsourcing/domain/model/aggregate.py]: Another excellent observation!

File: eventsourcing/types.py (hunk 1 above)
Comment (johnbywater) [reply eventsourcing/types.py]: Yes, that looks nicer. I actually changed it to look like this, then noticed mypy was complaining about redefinition of the metaclass. So I moved the 3.7 version up, and then there's the try/except with a "type: ignore" comment. Mypy seems happy with that. Will push changes....

Have also removed more of the "ignore-errors" in setup.cfg. And fixed up some more things. Mypy is green again now...

**Pull request merged**